literate_examples/example_cg.jl

```diff
@@ -0,0 +1,182 @@
+# # Conjugate Gradient
+# In this example we describe the basics of using the conjugate gradient iterative solvers
+# At the end you should be able to
+# 1. Use Conjugate Gradient to solve a linear system
+# 1. Know when to not use it
+# 1. Contruct a column-wise linear solver with Conjugate Gradient
+
+# ## What is it?
+# Conjugate Gradient is an iterative method for solving special kinds of linear systems:
+# ```math
+#  Ax = b
+# ```
+# via iterative methods.
+# !!! warning
+#     The linear operator need to be symmetric positive definite and the preconditioner must be symmetric.
+# See the [wikipedia](https://en.wikipedia.org/wiki/Conjugate_gradient_method) for more details.
+
+# ## Basic Example
+# First we must load a few things
+using CLIMA, CLIMA.LinearSolvers, CLIMA.ConjugateGradientSolver
+using LinearAlgebra, Random
+using GPUifyLoops, StaticArrays
+
+# Next we define a 3x3 symmetric positive definite linear system. (In the CliMa code a symmetric positive definite system could arise from treating diffusion implicitly.)
+const A = [
+    2.0 -1.0 0.0
+    -1.0 2.0 -1.0
+    0.0 -1.0 2.0
+];
+# We define the matrix `A` here as a global variable for convenience later.
+
+# We can see that it is symmetric. We can check that it is positive definite by checking the spectrum
+eigvals(A)
+
+# The linear operators that are passed into the abstract iterative solvers need to be defined as functions that act on vectors. Let us do that with our matrix.
+function linear_operator!(x, y)
+    mul!(x, A, y)
+    return nothing
+end;
+
+# We now define our `b` in the linear system
+b = ones(typeof(1.0), 3);
+
+# The exact solution to the system `Ax = b` is
+x_exact = [1.5, 2.0, 1.5];
+
+# Now we can set up the ConjugateGradient struct
+linearsolver = ConjugateGradient(b);
+# and an initial guess for the iterative solver.
+x = ones(typeof(1.0), 3);
+# To solve the linear system we just need to pass to the linearsolve! function
+iters = linearsolve!(linear_operator!, linearsolver, x, b)
+# The variable `x` gets overwitten during the linear solve
+# The norm of the error is
+norm(x - x_exact) / norm(x_exact)
+# The relative norm of the residual is
+norm(A * x - b) / norm(b)
+# The number of iterations is
+iters
+# Conjugate Gradient is guaranteed to converge in 3 iterations with perfect arithmetic in this case.
+
+# ## Non-Example
+
+# Conjugate Gradient is not guaranteed to converge with nonsymmetric matrices. Consider
+const A = [
```

Comment:
This gives
```
WARNING: redefining constant A
```

Reply:
I'll rewrite some of the functions so that I don't have to use constants. Ideally this will be something like
```julia
linearoperator!(x ,y) = mul!(x, A , y)
```